Various small fixes

kgaillot opened the pull request:
This is a collection of fixes for small issues I ran across while working on other stuff.

kgaillot commented:
This also correctly handles the issue addressed by #2127

kgaillot commented:
@clumens , can you review this when you get a chance? Low priority

clumens commented on lib/pacemaker/pcmk_sched_notif.c:
```diff
@@ -131,89 +131,95 @@ expand_node_list(GListPtr list, char **uname, char **metal)
+ */
 static void
-expand_list(GListPtr list, char **rsc_list, char **node_list)
+add_to_string_list(char **str_list, const char *s)
```
This would be worth turning into a pcmk__ function.  We're doing something similar in native_output_string in lib/pengine/native.c, and I'm sure there are other places too.  I just can't find them off the top of my head.  Even better would be a function that let you pass the separator character, since I'm sure we make comma separated lists this way all over the place too.

kgaillot replied:
LOL, looking into it, I see we already did :) pcmk__add_word()

I see potential uses for it in build_parameter_list(), crm_log_args(), expand_node_list(), and expand_list(). native_output_string() uses GString, so that would stay separate. We could just use GString everywhere rather than a custom function; not sure if that's better (#2126 is looking at GString for some other use cases).

Unfortunately the different cases use spaces differently. I'll have to see whether they can converge on a single approach or maybe add more args to the common function.

kgaillot replied:
create_level_registration_xml() looks to be the only additional possibility using comma instead of space

kgaillot commented:
@clumens, can you review again? I rebased on master but otherwise the only changes are in these commits:

Refactor: libcrmcommon: make some logging globals static
Refactor: libcrmcommon: simplify logging functions
Refactor: libcrmcommon: make pcmk__add_word() more flexible
Refactor: libstonithd: use new string building function
Refactor: controller: use pcmk__add_word() in build_parameter_list()
Refactor: scheduler: use pcmk__add_word() when generating notification values
Refactor: scheduler: rename notification name/value creating function
Refactor: scheduler: make clone notifications more efficient

kgaillot commented:
Well clearly I messed something up since it's core dumping, but that will have to wait until tomorrow

clumens commented on lib/common/strings.c:
```diff
@@ -599,24 +599,42 @@ crm_str_table_dup(GHashTable *old_table)
-pcmk__add_word(char *list, const char *word)
+void
+pcmk__add_separated_word(char **list, size_t *len, const char *word,
+                         const char *separator)
```
What happens if separator is NULL?

kgaillot replied:
Whoops :) What do you think: should it default to "" or " "?

clumens replied:
A space is probably what we'd want most of the time.  I doubt it'll ever come up given the wrapper, but still.

clumens commented on daemons/controld/controld_execd.c:
```diff
@@ -558,6 +555,10 @@ build_parameter_list(const lrmd_event_data_t *op,
         }
     }
 
+    if (list[0] != '\0') {
```
coverity caught your segfault right here - what if list is NULL?

kgaillot replied:
Yeah I was rushing to get something out for review before I left yesterday, not a good idea :)

I'll take some time to add unit tests for the add-word functions too

clumens commented on daemons/controld/controld_execd.c:
```diff
@@ -534,15 +533,13 @@ build_parameter_list(const lrmd_event_data_t *op,
-
-            // spaces before and after make parsing simpler
-            sprintf(list + start, " %s ", param->rap_name);
+            if (list[0] == '\0') {
```
Probably worth checking for NULL here, too.

kgaillot commented:
Updated. Changed commits: "make pcmk__add_word() more flexible", "use pcmk__add_word() in build_parameter_list()". New commit: "add unit tests for string-building functions".

clumens commented on lib/common/strings.c:
```diff
@@ -599,24 +599,52 @@ crm_str_table_dup(GHashTable *old_table)
 
-        list = pcmk__realloc(list, len + strlen(word) + 2); // 2 = space + EOS
-        sprintf(list + len, " %s", word);
+    if ((word == NULL) || (word[0] == '\0')) {
```
You can use pcmk__str_empty for this.

kgaillot replied:
Doh! I'll also update cocci, it checks for word[0] == 0 but not '\0'

kgaillot replied:
cocci is the root canal of development tools, useful but painful

clumens commented on lib/common/tests/strings/pcmk__add_word.c:
```diff
@@ -0,0 +1,98 @@
+
+    pcmk__add_word(&list, &list_len, "hello");
+    pcmk__add_word(&list, &list_len, "world");
+    g_assert_cmpint(strcmp(list, "hello world"), ==, 0);
```
You can use g_assert_cmpstr here (and elsewhere).

clumens commented on .gitignore:
```diff
@@ -206,6 +206,7 @@ pacemaker-[a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9]
 /lib/common/tests/flags/pcmk_all_flags_set
 /lib/common/tests/flags/pcmk_any_flags_set
 /lib/common/tests/operations/parse_op_key
+/lib/common/tests/strings/pcmk__add_word
```
Always nice to see more unit tests getting added.  We'll have good coverage one of these days.

kgaillot commented:
@clumens, added 5 commits on top, hopefully the last

kgaillot commented:
Added one more commit on top

clumens commented on lib/common/tests/operations/parse_op_key.c:
```diff
@@ -164,13 +164,13 @@ empty_input(void)
-    g_assert_true(rsc == NULL);
-    g_assert_true(ty == NULL);
+    g_assert_null(rsc);
+    g_assert_null(ty);
```
I didn't think we could use g_assert_null.  It was introduced in 2.38, and configure.ac has:
```
# Require glib 2.16.0 (2008-03) or later for g_hash_table_iter_init() etc.
PKG_CHECK_MODULES([GLIB], [glib-2.0 >= 2.16.0],
```

kgaillot replied:
Doh, didn't realize it was new :(

kgaillot replied:
Looks like just true/false/null/non-null are too new

kgaillot commented:
@clumens , I updated the "glib assertions" commit
